crates/runtime/src/externref.rs

```diff
@@ -393,6 +461,559 @@ impl Deref for VMExternRef {
+        //    TODO: One thing we *could* do to make this a little less error
+        //    prone, would be to assert/check that the nearest GC safepoint
+        //    found is within `max_encoded_size(any kind of call instruction)`
+        //    our queried PC for the target architecture.
+        //
+        // 2. Cranelift's stack maps only handle the stack, not
+        //    registers. However, some references that are arguments to a call
+        //    may need to be in registers. In these cases, what Cranelift will
+        //    do is:
+        //
+        //      a. spill all the live references,
+        //      b. insert a GC safepoint for those references,
+        //      c. reload the references into registers, and finally
+        //      d. make the call.
+        //
+        //    Step (c) adds drift between the GC safepoint and the location of
+        //    the call, which is where we actually walk the stack frame and
+        //    collect its live references.
+        //
+        //    Luckily, the spill stack slots for the live references are still
+        //    up to date, so we can still find all the on-stack roots.
+        //    Furthermore, we do not have a moving GC, so we don't need to worry
+        //    whether the following code will reuse the references in registers
+        //    (which would not have been updated to point to the moved objects)
+        //    or reload from the stack slots (which would have been updated to
+        //    point to the moved objects).
+        let index = match stack_maps
+            .pc_to_stack_map
+            .binary_search_by_key(&pc, |(pc, _stack_map)| *pc)
+        {
+            // Exact hit.
+            Ok(i) => i,
+
+            Err(n) => {
+                // `Err(0)` means that the associated stack map would have been
+                // the first element in the array if this pc had an associated
+                // stack map, but this pc does not have an associated stack
+                // map. That doesn't make sense since every call and trap inside
+                // Wasm is a GC safepoint and should have a stack map, and the
+                // only way to have Wasm frames under this native frame is if we
+                // are at a call or a trap.
+                debug_assert!(n != 0);
+
+                n - 1
+            }
+        };
+
+        let stack_map = stack_maps.pc_to_stack_map[index].1.clone();
+        Some(stack_map)
+    }
+}
+
+impl StackMapRegistryInner {
+    fn module_stack_maps(&self, pc: usize) -> Option<&ModuleStackMaps> {
+        let (end, stack_maps) = self.ranges.range(pc..).next()?;
+        if pc < stack_maps.range.start || *end < pc {
+            None
+        } else {
+            Some(stack_maps)
+        }
+    }
+}
+
+/// The registration for a module's stack maps.
+///
+/// Unsafe to drop earlier than its module is dropped. See
+/// `StackMapRegistry::register_stack_maps` for details.
+pub struct StackMapRegistration {
```

Comment:
Maybe panic in the drop implementation instead and add an unsafe `finish` function?